Allow ansible-galaxy to work offline

willthames opened the pull request:
##### ISSUE TYPE
- Bugfix Pull Request
##### ANSIBLE VERSION

```
ansible 2.1.0 (devel 5d69a09fd7) last updated 2016/04/26 13:15:28 (GMT +1000)
  lib/ansible/modules/core: (detached HEAD 4549ea5e85) last updated 2016/04/26 13:27:28 (GMT +1000)
  lib/ansible/modules/extras: (detached HEAD 86f08bfcda) last updated 2016/04/26 13:27:28 (GMT +1000)
  config file = 
  configured module search path = Default w/o overrides
```
##### SUMMARY

Allow ansible-galaxy to work offline

Remove all unnecessary connections to Galaxy API - only call the API
when required.

Fixes #14486
Fixes #13991

Alternative to #15363

willthames commented:
Rebasing on top of #15595 to ensure that tests actually pass.

bcoca commented:
you have extra commit here

bcoca commented on lib/ansible/cli/galaxy.py:
```diff
@@ -149,11 +149,6 @@ def run(self):
-        # if not offline, get connect to galaxy api
-        if self.action in ("import","info","install","search","login","setup","delete") or \
-            (self.action == 'init' and not self.options.offline):
-            self.api = GalaxyAPI(self.galaxy)
```
we cannot remove this as other options use it, but it might make sense to instantiate it inside those options to avoid the eternally growing list and cases here.

for example, search and all galaxy.ansible.com related actions break

willthames committed:
Allow galaxy to work offline

Remove all unnecessary connections to Galaxy API - only call the API
when required.

Fixes #14486
Fixes #13991

Alternative to #15363

willthames commented:
@bcoca you're quite right - not sure how I missed all those other `self.api`s

Latest revision should fix this. I'm not sure I prefer the decorated alternative - just seems simpler to create an api object if you need it and not bother if you don't.

bcoca commented:
I just see it as simpler going forward, as we add more options I did not want to keep adding more exceptions or anything else, it will just 'be there' if needed.

willthames commented:
Closing in favour of #15632
